src/sql/src/kafka_util.rs

```diff
@@ -47,10 +46,61 @@ generate_extracted_config!(
     ),
     (StatisticsIntervalMs, i32, Default(1_000)),
     (TopicMetadataRefreshIntervalMs, i32),
-    (TransactionTimeoutMs, i32)
+    (TransactionTimeoutMs, i32),
+    (StartTimeOffset, i64),
+    (StartOffset, KafkaStartOffset)
 );
 
-impl TryFrom<KafkaConfigOptionExtracted> for BTreeMap<String, StringOrSecret> {
+/// A fully specified lets of start offsets for a Kafka consumer.
+#[derive(Debug)]
+pub struct KafkaStartOffset(pub Vec<i64>);
+
+impl TryFromValue<Value> for KafkaStartOffset {
+    fn try_from_value(v: Value) -> Result<Self, PlanError> {
+        match v {
+            Value::Number(v) => Ok(Self(vec![v
+                .parse::<i64>()
+                .map_err(|e| sql_err!("invalid numeric value: {e}"))?])),
+            Value::Array(a) => {
+                let res: Result<Vec<i64>, PlanError> = a
+                    .into_iter()
+                    .map(|v| match v {
+                        Value::Number(v) => v
+                            .parse::<i64>()
+                            .map_err(|e| sql_err!("invalid numeric value: {e}")),
+                        _ => sql_bail!("cannot use value as number"),
+                    })
+                    .collect();
+
+                Ok(Self(res?))
+            }
+            _ => sql_bail!("cannot use value as number"),
+        }
+    }
+}
+
+impl ImpliedValue for KafkaStartOffset {
+    fn implied_value() -> Result<Self, PlanError> {
+        sql_bail!("must provide an array or number value")
+    }
+}
+
+/// An enum that represents start offsets for a consumer.
+/// `StartTimeOffset` should be eliminated after purification.
+#[derive(Debug)]
+pub enum KafkaStartOffsetType {
+    /// Fully specified offsets.
+    StartOffset(KafkaStartOffset),
+    /// A timestamp that will have to be translated
+    /// to offsets.
+    StartTimeOffset(i64),
+    /// Default behavior (start at 0).
+    None,
```

Comment:
This seems like where you have `KafkaStartOffsetType`, it just wants to be `Option<KafkaStartOffsetType>`; is there something that makes that onerous?

Reply:
thats definitely doable!